### PR TITLE
st1 [2183][ADD] stock_picking_partner_invoice_address

### PR DESCRIPTION
**Pull request by kakurai8**
[2183](https://www.quartile.co/web?#id=2183&action=771&model=project.task&view_type=form&menu_id=505)

- add partner_invoice_id to stock_picking form, tree
- add compute function of Boolean item wether partner_invoice_id = partner_shipping_id or not
- add filter wether partner_invoice_id = partner_shipping_id or not

### CODE REVIEW

File: stock_picking_filter/models/stock_picking.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Copyright 2019 C2i Change 2 improve - Eduardo Magdalena <emagdalena@c2i.es>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import fields, models
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    origin_invoiced_id = fields.Many2one(related="sale_id.partner_invoice_id")
```
Comment (yostashiro) [inline stock_picking_filter/models/stock_picking.py]:
```suggestion
    partner_invoice_id = fields.Many2one(related="sale_id.partner_invoice_id", store=True)
```

File: stock_picking_filter/readme/DESCRIPTION.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+This module does the following:
+- Adds partner_invoice_id field to stock picking.
```
Comment (yostashiro) [inline stock_picking_filter/readme/DESCRIPTION.rst]: Please add more background info or context wrt why this customization is needed.

File: stock_picking_filter/models/stock_picking.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Copyright 2019 C2i Change 2 improve - Eduardo Magdalena <emagdalena@c2i.es>
```
Comment (yostashiro) [inline stock_picking_filter/models/stock_picking.py]: Avoid copying a random module to create ours (even if in the draft stage).

**Review by yostashiro — CHANGES_REQUESTED** [review]
I'd name the module `stock_picking_partner_invoice_address` to be more explicit.

File: stock_picking_filter/models/stock_picking.py (hunk 3 above)
Comment (kakurai8) [reply stock_picking_filter/models/stock_picking.py]: I understand and fixed it.

File: stock_picking_filter/readme/DESCRIPTION.rst (hunk 2 above)
Comment (kakurai8) [reply stock_picking_filter/readme/DESCRIPTION.rst]: I've updated it but may adjust it further.

File: stock_picking_filter/models/stock_picking.py (hunk 1 above)
Comment (kakurai8) [reply stock_picking_filter/models/stock_picking.py]: I understand and fixed it.

File: stock_picking_partner_invoice_address/models/stock_picking.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from odoo import fields, models
+# from odoo import api, fields, models
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    partner_invoice_id = fields.Many2one(string="Invoice Address", related="sale_id.partner_invoice_id", store=True)
```
Comment (yostashiro) [inline stock_picking_partner_invoice_address/models/stock_picking.py]: Actually wasn't sure it behaves OK when sale_id is not available (e.g. receipt).  If not, we should use compute instead.

File: stock_picking_partner_invoice_address/models/stock_picking.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    same_invoiced_id = fields.Boolean(string="Partner = Invoice ", store=True, help="Whether or not the trading partner address is the same as the Invoice Address.")
```
Comment (yostashiro) [inline stock_picking_partner_invoice_address/models/stock_picking.py]: Avoid a field name ending with `_id` when it's not a many2one field. My preference: `is_identical_partners`

File: stock_picking_partner_invoice_address/views/stock_picking_view.xml [hunk 6]
```diff
@@ -0,0 +1,44 @@
+<odoo>
+    <record id="view_picking_form" model="ir.ui.view">
+        <field name="name">stock.picking.form</field>
+        <field name="model">stock.picking</field>
+        <field name="inherit_id" ref="stock.view_picking_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='partner_id']" position="after">
+                <field name="partner_invoice_id"/>
+                <field name="same_invoiced_id"/>
```
Comment (yostashiro) [inline stock_picking_partner_invoice_address/views/stock_picking_view.xml]: Show these fields only for outgoing pickings.

File: stock_picking_partner_invoice_address/views/stock_picking_view.xml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            </xpath>
+        </field>
+    </record>
+    <record id="vpicktree" model="ir.ui.view">
+        <field name="name">stock.picking.tree</field>
+        <field name="model">stock.picking</field>
+        <field name="inherit_id" ref="stock.vpicktree" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='partner_id']" position="after">
+                <field name="partner_invoice_id"/>
```
Comment (yostashiro) [inline stock_picking_partner_invoice_address/views/stock_picking_view.xml:25]: Note that this field will be showing regardless of the operation type.  If we want to adjust the tree view based on the operation type, we need to do something like this: https://github.dev/qrtl/hls-custom/blob/e7528ff083ddf7198e53f4160428768014ac103a/stock_picking_delivery_due_date/models/stock_picking.py#L33-L51

File: stock_picking_partner_invoice_address/models/stock_picking.py (hunk 5 above)
Comment (kakurai8) [reply stock_picking_partner_invoice_address/models/stock_picking.py]: I fixed this.

File: stock_picking_partner_invoice_address/views/stock_picking_view.xml (hunk 6 above)
Comment (kakurai8) [reply stock_picking_partner_invoice_address/views/stock_picking_view.xml]: I fixed this.

File: stock_picking_partner_invoice_address/views/stock_picking_view.xml (hunk 7 above)
Comment (kakurai8) [reply stock_picking_partner_invoice_address/views/stock_picking_view.xml:25]: This part is currently not working, so I have commented out.

File: stock_picking_partner_invoice_address/models/stock_picking.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        help="Whether or not the trading partner address is the same as the Invoice Address.",
+    )
+
+    @api.one
+    # test-compute 1
+    def _compute_is_identical_partners(self):
+        # Depending on whether the partner_invoiced_id is the same or not, the data to be linked to the shipping company and the template to be used for export will be different.
+        if not self.active or not self.is_identical_partners and self.partner_id:
+            return
+        stock_picking.is_identical_partners = True
```
Comment (yostashiro) [inline stock_picking_partner_invoice_address/models/stock_picking.py]:
```suggestion
    @api.multi
    @api.depends("partner_id", "partner_invoice_id")
    def _compute_is_identical_partners(self):
        for pick in self:
            if pick.partner_id == pick.partner_invoice_id:
                pick.is_identical_partners = True
```

**Comment by kakurai8** [comment]
> stock_quant_list_view/__manifest__.py

I wanted to exclude it from the update, but I deleted it by mistake. I will put it back in later in the week.

File: stock_picking_partner_invoice_address/readme/DESCRIPTION.rst [hunk 9]
```diff
@@ -0,0 +1,3 @@
+This module does the following:
+- Adds partner_invoice_id field to stock picking.
+- Depending on whether the partner_invoiced_id is the same or not, the data to be linked to the shipping company and the template to be used for export will be different.
```
Comment (yostashiro) [inline stock_picking_partner_invoice_address/readme/DESCRIPTION.rst]: > Depending on whether the partner_invoiced_id is the same or not, the data to be linked to the shipping company and the template to be used for export will be different.

This is not the function of the module. This is rather an expected usage which requires manual adjustments on the export layout.

File: stock_picking_partner_invoice_address/views/stock_picking_view.xml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        <field name="inherit_id" ref="stock.view_picking_internal_search" />
+        <field name="arch" type="xml">
+            <xpath expr="//filter[@name='draft']" position="before">
+                <filter
+                    string="Same as the Invoice Address"
+                    name="same"
+                    domain="[('is_identical_partners', '=', 'True')]"
+                />
+                <filter
+                    string="Different as the Invoice Address"
```
Comment (yostashiro) [inline stock_picking_partner_invoice_address/views/stock_picking_view.xml]:
```suggestion
                    string="Different to the Invoice Address"
```

File: stock_picking_partner_invoice_address/models/stock_picking.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2022 Quartile Limited
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from odoo import api, fields, models
+
+# from odoo import api, fields, models
```
Comment (yostashiro) [inline stock_picking_partner_invoice_address/models/stock_picking.py]:
```suggestion
```

**Review by yostashiro — APPROVED** [review]

**Pull request merged**